src/ngraph/runtime/gpu/gpu_cuda_kernel_builder.cpp

```diff
@@ -498,6 +501,286 @@ void runtime::gpu::CudaKernelBuilder::get_softmax_op(codegen::CodeWriter& writer
     return;
 }
 
+void runtime::gpu::CudaKernelBuilder::get_softmax_block_reduce_op(
+    codegen::CodeWriter& writer,
+    const std::string& name,
+    runtime::gpu::GPUKernelArgs& args,
+    const std::vector<std::string>& data_types,
+    size_t non_reduce_rank,
+    size_t reduce_rank,
+    size_t block_size_x)
+{
+    writer << runtime::gpu::nvrtc::helpers();
+    auto get_reduce_input_lambda = [&]() {
+        collective_coordinate_transform_helper(writer,
+                                               "reduce_idx",
+                                               "reduce_strides",
+                                               "reduce_strides_magic",
+                                               "reduce_strides_shift",
+                                               "reduce_strides_in_input",
+                                               "reduce_coordinate",
+                                               "reduce_input_index",
+                                               reduce_rank,
+                                               true);
+        writer << "input_idx = reduce_input_index + non_reduce_input_index;\n";
+        writer << "input_i = in[input_idx];\n";
+    };
+
+    auto stable_sum_lambda = [&]() {
+        writer << "input_i = expf(input_i - r_max);\n";
+        writer << "y = input_i - c;\n";
+        writer << "t = r_sum + y;\n";
+        writer << "c = (t - r_sum) - y;\n";
+        writer << "r_sum = t;\n";
+    };
+
+    auto max_lambda = [&]() { writer << "r_max = r_max > input_i ? r_max : input_i;\n"; };
+
+    auto divide_lambda = [&]() {
+        writer << "input_i = expf(input_i - r_max) / r_sum;\n";
+        writer << "out[input_idx] = input_i;\n";
+    };
+
+    // auto unroll_loop_lambda = [&](void (*func)()) -> void {
```

Comment:
Can we remove this commented out code or is it to be used later?

Reply:
yes, removed for now. Might add later.